### ISSUE TITLE
Naming model outputs by layer causes collision when sharing layer among different output tensors

### ISSUE THREAD

**Issue by mogeid** [keras-team/keras#14031]
When a model is created, the `output_names` property inherited from Network takes those names from the output layer(s). When the outputs are produced by different layers, there is no issue in using multiple losses or a multi-output loss as these can be matched by passing a loss dictionary when compiling the model:
```python
model.compile(optimizer=opt, loss={'output_layer_1': loss_fn1, 'output_layer_2':loss_fn2})
# or
model.compile(optimizer=opt, loss={'output_layer_1': loss_fn, 'output_layer_2':loss_fn})
```
However, when sharing a layer for multiple outputs, there is no way to identify these tensors as they come from the same layer and would have the same key:
```python
layer = SomeLayer(args, name='output_layer')
output1 = layer(tensor1)
output2 = layer(tensor2)
model = Model(..., [output1, output2])
print(model.output_names)
# : ['output_layer', 'output_layer']
model.compile(optimizer=opt, loss={'output_layer': loss_fn1, 'output_layer':loss_fn2})  # improper syntax
```
https://github.com/keras-team/keras/blob/7a39b6c62d43c25472b2c2476bd2a8983ae4f682/keras/engine/network.py#L300

**Comment by mogeid** [comment]
I suppose you can get around this by using separate identity Lambda layers for each output, but still.

**Comment by Saduf2019** [comment]
@mogeid 
hey we were supposed to connect for that dataset, did u go tru it?
